### PR TITLE
feat(Messaging): Adds Kafka support

### LINKED ISSUE #8591: Adding Kafka support to Hermes
We use Rucio in our project (Rubin Observatory), and have modified Hermes to include Kafka support.   This works in much the same way that ActiveMQ in Hermes works.

Changes are to the rucio.cfg file to include a new section [messaging-hermes-kafka], the addition of KAFKA to the constants.py file for the HermesServes class, and the code additions to hermes.py.

Adding this to the codebase would help us out because we've been modifying the Hermes code each time we upgrade to a new version of Rucio.

I've got a PR just about ready to go, and I'm going through the contributing guidelines for Rucio so I'll be able to submit it properly.

### PR DESCRIPTION
**Pull request by srp3rd**
Closes: #8591

*By submitting this PR, I confirm I have followed the [Contributing Guide](https://rucio.cern.ch/documentation/contributing/).*

## Checklist
- [x] Created issue: #8591
- [x] Added tests
- [x] Updated documentation (Please link PRs in documentation repository) https://github.com/rucio/documentation/pull/844
- [ ] Added database migrations (if needed)
- [ ] For backwards compatibility breaking changes, leave additional description, follow conventional commits
- [ ] Picked a reviewer after submission (Leave empty, if unclear)
- [x] I understand that stale (failing tests, author not answering) PRs will be closed promptly

## Additional description for reviewer

*Note: This OPTIONAL is only relevant for the REVIEWER, please leave it in the PR*

<details>
<summary>Reviewer template </summary>
Reviewers should copy&paste the code-block below and fill it out for APPROVED pull requests.
If the PR does not meet the standards the project sets out, the reasons should be WELL EXPLAINED in a CHANGE REQUEST (The answers below do not need to be answered in that case)

- **Confidence in review**: I am confident in my review concerning the components this PR touches: [*High 🟢, Medium 🟡 Low 🔴*]
- **Confidence in scope**: I am confident that this fits into the scope of the project and should be included: [*High 🟢, Medium 🟡, Low 🔴*]
  - For Medium and Low, explain in notes why this should be included
- **Quality**: The approach is sound, maintainable and addresses the issue in the best way: [*Agree 🟢*]
- **Security**: This PR does NOT require increased attention in terms of security (E.g. new dependencies): [*Agree 🟢, Disagree 🔴*]
  - If *Disagree* explain in notes.
- **Backwards compatibility**: This PR does NOT introduce backwards compatibility breaking changes: [*Agree 🟢, Disagree 🔴*]
  - If *Disagree* explain in notes
- **Testing**: This PR is well tested: [*Agree 🟢*]
- **Documentation**: Relevant documentation or comments are updated or not required: [*Agree 🟢*]

```
- **Confidence in review**: High 🟢 Medium 🟡 Low 🔴
- **Confidence in scope**: High 🟢 Medium 🟡 Low 🔴
- **Quality**: Agree 🟢
- **Security**: Agree 🟢 Disagree 🔴
- **Backwards compatibility**: Agree 🟢 Disagree 🔴
- **Testing**: Agree 🟢
- **Documentation**: Agree 🟢

# Notes for merger



```
</details>

### CODE REVIEW

File: lib/rucio/daemons/hermes/hermes.py [hunk 1]
```diff
@@ -96,6 +97,76 @@ def on_error(self, frame: "Frame") -> None:
...
+        password = config_get("messaging-hermes-kafka", "password", raise_exception=False, default=None)
+
+    config = {'bootstrap.servers': f'{brokers}',
+               'client.id': socket.gethostname(),
+             }
+
+    # configure to use SSL
+    if use_ssl:
+        logger(logging.INFO, "[broker-kafka] use_ssl")
+        producer = Producer(bootstrap_servers=f'{brokers}',
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: Both this and the branch below (line 154) calling the `Producer` constructors raise errors at runtime as per their [docs](https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#pythonclient-producer). Keyword arguments are treated as literal config properties (dot-separated, e.g. `bootstrap.servers`), and the names given here are not valid properties, but they are for `kafka-python`'s `KafkaProducer`. I reproduced this locally:

Image

Which works with a dict:

Image

Do note, that the mapping is not 1 to 1 `_` to `.`, the ssl location for example is different.

File: lib/rucio/daemons/hermes/hermes.py [hunk 2]
```diff
@@ -96,6 +97,76 @@ def on_error(self, frame: "Frame") -> None:
...
+    # check to see if ssl is being used to authenticate
+    logger(logging.INFO, "[broker] Checking authentication method")
+    try:
+        use_ssl = config_get_bool("messaging-hermes-kafka", "use_ssl")
+    except Exception:
+        logger(logging.INFO, "[broker] Could not find use_ssl in configuration -- update your rucio.cfg")
+
+    # if ssl is used, get the reset of the params we'll use to authenticate
+    if use_ssl:
+        ca_cert = config_get_bool("messaging-hermes-kafka", "ca_cert")
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: These 3 properties are file paths, not booleans so `config_get_bool` will raise a value error here for anything that isn't false/true. These should be `config_get`

File: lib/rucio/daemons/hermes/hermes.py [hunk 3]
```diff
@@ -201,6 +272,34 @@ def setup_activemq(
...
+        try:
+            d = {
+                "event_type": str(message["event_type"]).lower(),
+                "payload": message["payload"],
+                "created_at": str(message["created_at"])
+            }
+
+            value = json.dumps(d)
+            logger(logging.DEBUG, "kafka sending to topic: %s -> %s", topic, value)
+            producer.produce(topic, key="message", value=value)
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: This is an asynchronous operation as per the [docs](https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#pythonclient-producer) and the `flush()` that happens on the next line does block until the queue drain, but if a failure happens, it will be silently discarded and `flush` will return normally. The message id is appended to `to_delete` regardless, so during any broker outage Hermes will delete the messages from the database that were never delivered.
The fix here would be an `on_delivery` callback that records only the successful ids.

File: lib/rucio/daemons/hermes/hermes.py [hunk 4]
```diff
@@ -96,6 +97,76 @@ def on_error(self, frame: "Frame") -> None:
...
+
+    # check to see if ssl is being used to authenticate
+    logger(logging.INFO, "[broker] Checking authentication method")
+    try:
+        use_ssl = config_get_bool("messaging-hermes-kafka", "use_ssl")
+    except Exception:
+        logger(logging.INFO, "[broker] Could not find use_ssl in configuration -- update your rucio.cfg")
+
+    # if ssl is used, get the reset of the params we'll use to authenticate
+    if use_ssl:
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py:138]: If the above part enters the exception at line 128, it will result in an unbound `use_ssl` so it needs a default value. You can skip the try-except error handling by using the `config_get_bool` args `raise_exception=False, default=False`

File: lib/rucio/daemons/hermes/hermes.py [hunk 5]
```diff
@@ -201,6 +272,34 @@ def setup_activemq(
...
+            d = {
+                "event_type": str(message["event_type"]).lower(),
+                "payload": message["payload"],
+                "created_at": str(message["created_at"])
+            }
+
+            value = json.dumps(d)
+            logger(logging.DEBUG, "kafka sending to topic: %s -> %s", topic, value)
+            producer.produce(topic, key="message", value=value)
+            producer.flush()
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: Flushing here for every message could be pretty tough on the thread as, if the broker is unreachable, it will result in waiting a lot per message. Optionally we could use the `timeout` parameter for the flush and set it really small but I still think we should move it outside of the for-loop.
I think the default value for it is here https://docs.confluent.io/platform/current/clients/librdkafka/html/md_CONFIGURATION.html called `message.timeout.ms` which defaults to 300 seconds.

File: lib/rucio/daemons/hermes/hermes.py [hunk 6]
```diff
@@ -36,6 +36,7 @@
 
 import requests
 import stomp
+from confluent_kafka import Producer
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: This import should be lazy imported via `rucio.common.extra.import_extras` like the `globus_sdk` import, for example
```python
EXTRA_MODULES = import_extras(['confluent_kafka'])                                                                                                                                                                                                                             
if EXTRA_MODULES['confluent_kafka']:                                                                                                                                                                                                                                           
     from confluent_kafka import Producer # pylint: disable=import-error
```

File: requirements/requirements.server.txt [hunk 7]
```diff
@@ -47,6 +47,8 @@ charset-normalizer==3.4.7
     # via requests
 click==8.1.8
     # via flask
+confluent-kafka==2.14.2
```
Comment (VladGaluska) [inline requirements/requirements.server.txt:50]: This also needs to go into `pyproject.server.toml` as `kafka = ['confluent-kafka<=2.14.2']`

File: requirements/requirements.dev.in [hunk 8]
```diff
@@ -16,3 +16,5 @@ pip-tools==7.5.3                                            # Generate requireme
 atlas-rucio-policy-package==0.8.3
 belleii-rucio-policy-package @ git+https://github.com/belle2/Belle-II-Rucio-Policy-Package.git@v0.1.1
 ruff==0.15.20                                               # Linter
+confluent_kafka==2.14.2                                     # Confluent Kafka
+retrying==1.4.2                                             # Adds retry behavior - used by Confluent Kafka
```
Comment (VladGaluska) [inline requirements/requirements.dev.in]: We can remove this as `confluent_kafka` only requires `typing-extensions`:
```shell
pip show confluent-kafka | grep Requires 
Requires: typing-extensions
```

File: requirements/requirements.dev.txt
Comment (VladGaluska) [inline requirements/requirements.dev.txt:1]: Looks like the requirements here were regenerated inside a container. It works fine, but it's inconsistent and results in a huge change file. Could you please regenerate this from inside the `requirements/` directory so only the added packages show up in the diff? Thanks! Also for `requirements.server.in`

File: lib/rucio/daemons/hermes/hermes.py [hunk 9]
```diff
@@ -901,6 +1005,26 @@ def run_once(heartbeat_handler: "HeartbeatHandler", bulk: int, **_kwargs) -> boo
                         to_delete.append(message)
             except Exception as error:
                 logger(logging.ERROR, "Error sending to syslog : %s", str(error))
+        if "kafka" in message_dict:
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: This needs an extra `if "kafka" in services_list` check, in the off chance that `kafka` is later removed from `services_list` while kafka-tagged messages are still in the table.

File: lib/rucio/daemons/hermes/hermes.py [hunk 10]
```diff
@@ -755,6 +854,11 @@ def run_once(heartbeat_handler: "HeartbeatHandler", bulk: int, **_kwargs) -> boo
             logger(logging.ERROR, str(err))
     if "syslog" in services_list:
         logger(logging.INFO, "Syslog service enabled")
+    if "kafka" in services_list:
+        try:
+            producer, topic = setup_kafka(logger)
+        except Exception as err:
+            logging.exception(err)
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: This is a bit of a nitpick, but every other "sibling" branch uses the logger here (including this function). Can you switch to `logger`?

File: lib/rucio/daemons/hermes/hermes.py [hunk 11]
```diff
@@ -96,6 +97,76 @@ def on_error(self, frame: "Frame") -> None:
...
+
+    logger(logging.INFO, "[broker-kafka] Resolving brokers")
+
+    # the following retrieves all rucio.cfg information
+
+    # get broker name
+    try:
+        brokers = config_get("messaging-hermes-kafka", "brokers")
+    except Exception:
+        raise Exception("Could not load 'brokers' from configuration")
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: This swallows the original exception to produce an exception of the same type (and not a Rucio exception type). We should at least log the original exception and propagate a rucio exception type.

File: lib/rucio/daemons/hermes/hermes.py [hunk 12]
```diff
@@ -96,6 +97,76 @@ def on_error(self, frame: "Frame") -> None:
...
+        producer = Producer(bootstrap_servers=f'{brokers}',
+                                 sasl_username=username,
+                                 sasl_password=password)
+    else:
+        logger(logging.INFO, f"[broker-kafka] plain {config}")
+        producer = Producer(config)
+
+    # retrieve the topic
+    try:
+        topic = config_get("messaging-hermes-kafka", "topic", raise_exception=True, default=None)
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: `default=None` here is dead code because of `raise_exception`

File: lib/rucio/daemons/hermes/hermes.py [hunk 13]
```diff
@@ -96,6 +97,76 @@ def on_error(self, frame: "Frame") -> None:
...
+    except Exception:
+        raise Exception("Could not load kafka topic from configuration")
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: Same as the previous `raise Exception`, it should log and switch to a better exception type.

File: lib/rucio/daemons/hermes/hermes.py (hunk 3 above)
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: Using the same key for all the message makes the default partitioner hash them in a single partition. For the `ActiveMQ` path, it sends no key equivalent so the correct way here would be to omit the `key` and let the messages be distributed randomly. If ordering matters here, we should use a meaningful key (like scope or event type), but a constant key doesn't do much.

File: etc/docker/test/extra/rucio_default.cfg [hunk 14]
```diff
@@ -171,3 +171,9 @@ gcs = /opt/rucio/etc/google-cloud-storage-test.json
 
 [api]
 endpoints = accountlimits, accounts, archives, auth, config, credentials, dids, dirac, export, heartbeats, identities, import, lifetime_exceptions, locks, meta_conventions, ping, redirect, replicas, requests, rses, rules, scopes, subscriptions, traces, vos, opendata, opendata_public
+
+[messaging-hermes-kafka]
+nonssl_port = 9092
```
Comment (VladGaluska) [inline etc/docker/test/extra/rucio_default.cfg]: This doesn't appear to be used

File: etc/docker/test/extra/rucio_default.cfg [hunk 15]
```diff
@@ -171,3 +171,9 @@ gcs = /opt/rucio/etc/google-cloud-storage-test.json
...
+use_ssl = False
+brokers = kafka
```
Comment (VladGaluska) [inline etc/docker/test/extra/rucio_default.cfg:177]: There is no kafka service in the docker-compose.yml so this would not do anything for the dev environment (which is the place that this config file is used, CI doesn't use it)

File: lib/rucio/daemons/hermes/hermes.py [hunk 16]
```diff
@@ -755,6 +854,11 @@ def run_once(heartbeat_handler: "HeartbeatHandler", bulk: int, **_kwargs) -> boo
             logger(logging.ERROR, str(err))
     if "syslog" in services_list:
         logger(logging.INFO, "Syslog service enabled")
+    if "kafka" in services_list:
+        try:
+            producer, topic = setup_kafka(logger)
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: This recomputes the `Producer` on every run_once cycle (every ~10s default). Kafka producers should to be long-lived so this should be set up once and reused across cycles. `ActiveMQ` constructs it the same way, but I think it's more expensive for kafka: it spins up background threads and immediately opens broker connections/metadata fetches on construction.

File: tests/test_hermes.py [hunk 17]
```diff
@@ -434,3 +434,74 @@ def test_hermes_syslog_tcp(core_config_mock, caches_mock):
...
+    [
+        {
+            "caches_to_mock": [
+                "rucio.core.config.REGION",
+            ]
+        }
+    ],
+    indirect=True,
+)
+def test_hermes_kafka(core_config_mock, caches_mock):
```
Comment (VladGaluska) [inline tests/test_hermes.py:465]: Hmm, the existing tests in `test_hermes` consume from a real ActiveMQ container in the dev compose and asserts that the messages were actually delivered. It does mock some things, but it's what the stack doesn't provide (SMTP I think). Could you add a corresponding kafka service to `etc/docker/dev/docker-compose.yml` and write an end to end test that consumes what Hermes pushes?
I don't know how urgent this PR should be merged, as this would increase it quite a bit. 
To be honest, just one end to end test would be the minimum, for me, as I see a lot of other test cases this should cover.

 Unit tests:
- config branch assertions since there were issues in this pr with it
- missing brokers/topic branch
- failed delivery -> messages are not deleted
- partial failure -> e.g. 2 messages succeed (and are deleted) and 1 doesn't (and will be queued again)
- `produce` raises 3 errors, so tests handling those would be good

Integration tests:
- no ssl happy path
- ssl happy path
- unreachable broker

**Review by VladGaluska — COMMENTED** [review]
Hi, thanks for doing this! :) It's a nice addition and you got the hard parts done. I left quite a few comments here, but most of them collapse into 3 fixes:
- producer construction
- delivery semantics
- packaging

Most of the comments will prob. fold into a single change. The test comments are the ones I find most important though, as the current test can't catch the constructor issue because the mock accepts anything.

Additionally, I think you're missing from `AUTHORS.rst` so please add yourself there also :)

**Comment by srp3rd** [comment]
I have addressed nearly all the issues locally and I'm running regression tests locally to be sure I didn't screw anything up.   I'm working on the docker-compose.yml/unit test issue next.  I expect to be AFK for the next couple of weeks, and will address that when I get back in front of a keyboard. I saw there was a three working day close out of tickets and don't want that to happen.

**Comment by VladGaluska** [comment]
Woah that's fast :D I wouldn't worry too much about the stale policy since there's active work being done here, and I've had a PR get auto-closed before that just got reopened and nothing got lost. Enjoy your time off!

File: lib/rucio/daemons/hermes/hermes.py (hunk 1 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: Fixed

File: lib/rucio/daemons/hermes/hermes.py (hunk 2 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: Fixed

File: lib/rucio/daemons/hermes/hermes.py (hunk 3 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: Fixed

File: lib/rucio/daemons/hermes/hermes.py (hunk 4 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py:138]: Fixed

File: lib/rucio/daemons/hermes/hermes.py (hunk 5 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: Now uses a default value of 300 seconds, and this can be adjusted in the `[messaging-hermes-kafka]` section by specifying the value in `message_timeout_ms`.

File: lib/rucio/daemons/hermes/hermes.py (hunk 6 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: Fixed as above

File: requirements/requirements.server.txt (hunk 7 above)
Comment (srp3rd) [reply requirements/requirements.server.txt:50]: added

File: requirements/requirements.dev.in (hunk 8 above)
Comment (srp3rd) [reply requirements/requirements.dev.in]: Doh....a previous version of the code used that.  I removed it.

File: lib/rucio/daemons/hermes/hermes.py (hunk 9 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: Added.  I also added that for the `activemq` section.

File: lib/rucio/daemons/hermes/hermes.py (hunk 10 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: switched to using `logger`

File: lib/rucio/daemons/hermes/hermes.py (hunk 11 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: Added the log message; I wasn't sure which Rucio Exception to use here, so I used ConfigurationError - should this be something else?

File: lib/rucio/daemons/hermes/hermes.py (hunk 12 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: Removed

File: lib/rucio/daemons/hermes/hermes.py (hunk 13 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: Changed to output a log message and raise ConfigurationError

File: lib/rucio/daemons/hermes/hermes.py (hunk 3 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: Removed `key=message`

File: etc/docker/test/extra/rucio_default.cfg (hunk 14 above)
Comment (srp3rd) [reply etc/docker/test/extra/rucio_default.cfg]: Removed

File: lib/rucio/daemons/hermes/hermes.py (hunk 16 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: I'm not sure how you wanted to handle this long term.   But here's what I came up with:

This was a little tricky because of the way things are executed in run_daemon.  I added a `longterm_state` dictionary that adds the output from `setup_kafka` to it, and it checks for the existence of those values the next time through.  Additional things that have nothing to do with kafka can go into `longterm_state`, if there's a need for that in the future.

**Comment by srp3rd** [comment]
A pushed the changes I've noted above;  I'm not done testing these yet.  I'll address any CI issues and the `docker-compose` work when I return from break.

File: lib/rucio/daemons/hermes/hermes.py (hunk 1 above)
Comment (VladGaluska) [reply lib/rucio/daemons/hermes/hermes.py]: I believe the `ssl` properties are still not valid [librdkafka](https://docs.confluent.io/platform/current/clients/librdkafka/html/md_CONFIGURATION.html) properties. For example, `ssl.cafile` -> `ssl.ca.location`

File: lib/rucio/daemons/hermes/hermes.py (hunk 5 above)
Comment (VladGaluska) [reply lib/rucio/daemons/hermes/hermes.py]: This should also then be passed as `'message.timeout.ms': message_timeout_ms` to the Producer configs

File: requirements/requirements.server.txt (hunk 7 above)
Comment (VladGaluska) [reply requirements/requirements.server.txt:50]: Ah that got added into `pyproject.toml` which is fine, but it's needed in `pyproject.server.toml`

File: lib/rucio/daemons/hermes/hermes.py (hunk 6 above)
Comment (VladGaluska) [reply lib/rucio/daemons/hermes/hermes.py]: Since the `Producer` is still referenced in type annotations these should have the `"` surrounding them, so as to not break, e.g. `Optional[Producer]` -> `Optional["Producer"]` . Then it will also have to be included in type checking, e.g.
```python
if TYPE_CHECKING:                                                                                                                                                                                                                                                                                
      from confluent_kafka import Producer
```

Additionally, while checking this existing pattern in other modules, I realised they also have an extra runtime guard that should be added at the top of `setup_kafka` :
```python
  if not EXTRA_MODULES['confluent_kafka']:                                                                                                                                                                                                                                                         
      raise ConfigurationError('kafka is configured in services_list but confluent-kafka is not installed')
```

Sorry about that, I missed those in my comment

File: lib/rucio/daemons/hermes/hermes.py [hunk 18]
```diff
@@ -201,6 +282,41 @@ def setup_activemq(
...
+    to_delete = []
+    msg_count = 0
+
+    def _on_delivery(err, kafka_msg, *, message_id):
+        nonlocal msg_count
+        if err is None:
+            to_delete.append(message_id)
+            msg_count += 1
+        else:
+            logger(logging.WARNING, "kafka delivery failed for %s: %s", message_id, str(err))
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: A bit of a nitpick, but to match line 314, this should use `logging.WARN`

File: lib/rucio/daemons/hermes/hermes.py (hunk 3 above)
Comment (VladGaluska) [reply lib/rucio/daemons/hermes/hermes.py]: Nice!

File: requirements/requirements.dev.in (hunk 8 above)
Comment (VladGaluska) [reply requirements/requirements.dev.in]: Looks like it's still in `requirements/requirements.server.in`

File: lib/rucio/daemons/hermes/hermes.py (hunk 11 above)
Comment (VladGaluska) [reply lib/rucio/daemons/hermes/hermes.py]: That looks good to me, but it would be better to propagate the error instead of using the message, e.g. `raise ConfigurationError(...) from err`

**Comment by VladGaluska** [comment]
I've taken a look at some of the comments, but I can't resolve them due to being an external contributor :( So I left a 👍 on them instead

File: lib/rucio/daemons/hermes/hermes.py (hunk 16 above)
Comment (VladGaluska) [reply lib/rucio/daemons/hermes/hermes.py]: That's neat (and elegant). But I now see a new failure mode that could happen. If flush does not send all the messages to kafka (because the broker is slower, or more data got sent than usual), they will still be queued for sending, and on another run of `run_once` it will resend that message (possibly the same message arrives multiple times to the client). It's not that bad, but we could fix it by keeping track of what's currently being sent to kafka via the new `longterm_state`, as well as keeping the `delivered_ids` there

**Comment by VladGaluska** [comment]
We can ping one of the maintainers when you're back and the PR is ready to reopen this @srp3rd  :)

**Comment by srp3rd** [comment]
> We can ping one of the maintainers when you're back and the PR is ready to reopen this @srp3rd :)

I've been working on this and have made a few comments about a pending push.  What's the best way to get this reopened?

File: lib/rucio/daemons/hermes/hermes.py (hunk 18 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: I checked into this, and the logging.WARN has been deprecated in favor of logging.WARNING, so I changed the logging.WARN.  This also make it consistent with the rest of hermes.py, which using logging.WARNING throughout

File: lib/rucio/daemons/hermes/hermes.py (hunk 5 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: Fixed - and removed the argument from the flush() call, because it's redundant.

File: lib/rucio/daemons/hermes/hermes.py (hunk 1 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: All of these have been fixed appropriately, and the code updated to remove duplicate code.

**Comment by VladGaluska** [comment]
We can tag one of the maintainers, maybe @voetberg ? Thank you in advance!

File: lib/rucio/daemons/hermes/hermes.py (hunk 16 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: Yeah, this came up because the Producer wasn't being recreated every time.  I added some code to keep track of this.

File: lib/rucio/daemons/hermes/hermes.py (hunk 6 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: Fixed

File: requirements/requirements.server.txt (hunk 7 above)
Comment (srp3rd) [reply requirements/requirements.server.txt:50]: Added

File: requirements/requirements.dev.txt
Comment (srp3rd) [reply requirements/requirements.dev.txt:1]: This has been addressed

File: requirements/requirements.dev.in (hunk 8 above)
Comment (srp3rd) [reply requirements/requirements.dev.in]: Fixed

File: lib/rucio/daemons/hermes/hermes.py (hunk 11 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: Fixed

File: etc/docker/test/extra/rucio_default.cfg (hunk 15 above)
Comment (srp3rd) [reply etc/docker/test/extra/rucio_default.cfg]: Removed

File: tests/test_hermes.py (hunk 17 above)
Comment (srp3rd) [reply tests/test_hermes.py:692]: I added the Kafka service and Zookeeper to `etc/docker/dev/docker-compose.yml` and reworked the unit test.  I think that covers it, but there is a mock there so we can still run the unit tests in an isolated way.

File: lib/rucio/daemons/hermes/hermes.py [hunk 19]
```diff
@@ -96,12 +133,95 @@ def on_error(self, frame: "Frame") -> None:
...
+                  'security.protocol': 'SSL',
+                  'ssl.ca.location': ca_cert,
+                  'ssl.certificate.location': certfile,
+                  'ssl.key.location':  keyfile,
+             })
+    elif username is not None:
+        logger(logging.INFO, "[broker-kafka] username")
+        config.update({
+                  'security.protocol': 'SASL_SSL',
+                  'security.mechanism': 'PLAIN',
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: I believe this should be `sasl.mechanism` as per the [config docs](https://docs.confluent.io/platform/current/clients/librdkafka/html/md_CONFIGURATION.html)

File: lib/rucio/daemons/hermes/hermes.py [hunk 20]
```diff
@@ -135,53 +255,117 @@ def setup_activemq(
...
+        try:
+            d = {
+                "event_type": str(message["event_type"]).lower(),
+                "payload": message["payload"],
+                "created_at": str(message["created_at"])
+            }
+
+            value = json.dumps(d)
+        except (TypeError, ValueError) as e:
+            logger(logging.ERROR, "can't serialize %s: %s", mid, str(e))
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py:343]: I believe the except block needs a continue here, because otherwise value is undefined and it will keep going in the for block

File: lib/rucio/daemons/hermes/hermes.py [hunk 21]
```diff
@@ -96,12 +133,95 @@ def on_error(self, frame: "Frame") -> None:
...
+    logger(logging.INFO, "[broker-kafka] Resolving brokers")
+
+    # the following retrieves all rucio.cfg information
+
+    # get broker name
+    try:
+        brokers = config_get("messaging-hermes-kafka", "brokers")
+    except Exception as err:
+        logger(logging.ERROR, "Could not load 'brokers' from configuration")
+        raise ConfigurationError(err)
```
Comment (VladGaluska) [inline lib/rucio/daemons/hermes/hermes.py]: Also needs a `from err` to not lose the traceback

File: lib/rucio/daemons/hermes/hermes.py (hunk 18 above)
Comment (VladGaluska) [reply lib/rucio/daemons/hermes/hermes.py]: Oh, that's good to know!

File: etc/docker/dev/docker-compose.yml [hunk 22]
```diff
@@ -559,6 +559,39 @@ services:
       iam-db:
         condition: service_healthy
 
+  zookeeper:
```
Comment (VladGaluska) [inline etc/docker/dev/docker-compose.yml:562]: This is just a suggestion, but all the other optional services here are behind a profile. Could you also add a `profiles: [kafka]` please? Also, I think `confluentinc/cp-server` is the enterprise version, which works as per their license agreement (they allow up to 1 broker), but it's a bit odd for a dev stack, and I don't think those extra features would be needed. It also comes with a hefty env, and at least the `KAFKA_CONFLUENT_SCHEMA_REGISTRY_URL` points to a url that doesn't exist.

It would be easier to use `docker.io/apache/kafka` as that also doesn't require zookeeper, e.g.:

```yml
  kafka:
    container_name: ${RUCIO_KAFKA_CONTAINER_NAME:-dev-kafka-1}
    image: docker.io/apache/kafka:4.3.1
    profiles:
      - kafka
    environment:
      KAFKA_NODE_ID: 1
      KAFKA_PROCESS_ROLES: broker,controller
      KAFKA_CONTROLLER_QUORUM_VOTERS: 1@localhost:9093
      KAFKA_CONTROLLER_LISTENER_NAMES: CONTROLLER
      KAFKA_LISTENERS: PLAINTEXT://:9092,CONTROLLER://:9093,PLAINTEXT_HOST://:29092
      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://kafka:9092,PLAINTEXT_HOST://localhost:29092
      KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT,PLAINTEXT_HOST:PLAINTEXT
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1
    healthcheck:
      test: ["CMD", "/opt/kafka/bin/kafka-broker-api-versions.sh", "--bootstrap-server", "localhost:9092"]
      interval: 5s
      timeout: 10s
      retries: 12
```
and then in `docker-compose.ports.yml`:
```yml
  kafka:
    ports:
      - "127.0.0.1:29092:29092"
```

^ tried this one out and it worked

File: tests/test_hermes.py (hunk 17 above)
Comment (VladGaluska) [reply tests/test_hermes.py:692]: Nice! Container and unit tests are there. There's still one part that's still missing, since, as you said, all the tests are using mocks, and none actually talk to kafka. For activemq there is an equivalent integration test `test_hermes` where it adds a few messages and runs `hermes.hermes(once=True) ` with the real producer. The kafka version should do the same with a real Producer and a real Consumer reading from `kafka:9092`, then asserting the payloads match and the rows are deleted. A test like this would be the only test that actually checks that the `librdkafka` config connects, which is also where we caught some bugs. You can also reuse the existing `core_config_mock` and just replace the brokers to say `kafka:9092`. Since the container will be optional if/when you add the profile, it will also need something like is done with `@skip_missing_elasticsearch_influxdb_in_env` .

Also, now that the container exists, the config we talked about earlier would make sense so Hermes can be run against it (sorry about that). It should go in `rucio_default.cfg` and `rucio_autotests_common.cfg`
 since these are both chosen depending on the RDBMS. Would have to look something like:
```
[hermes]
services_list = activemq,kafka

[messaging-hermes-kafka]
brokers = kafka:9092
topic = hermeskafka
use_ssl = False
```

**Review by VladGaluska — CHANGES_REQUESTED** [review]
Great changes, just a few more things left. Sorry for the late review, I was on holiday :D 

Two of the new comments are actually bugs and the rest are small follow-ups.

As before I left a 👍 on the threads I would've resolved.

File: lib/rucio/daemons/hermes/hermes.py (hunk 19 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: fixed

File: lib/rucio/daemons/hermes/hermes.py (hunk 20 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py:343]: you're right, of course - fixed

File: lib/rucio/daemons/hermes/hermes.py (hunk 21 above)
Comment (srp3rd) [reply lib/rucio/daemons/hermes/hermes.py]: fixed

File: tests/test_hermes.py (hunk 17 above)
Comment (srp3rd) [reply tests/test_hermes.py:692]: I'll look into this

Comment (srp3rd) [reply tests/test_hermes.py]: If I add the profile `kafka` to the `docker-compose.yml` file, it won't start up as part of the tests, unless I add `--profile, "kafka"` to `run_tests.py`.  However, I think that has other ramifications, because by adding that, it'll always start the kafka container for everything.  At least I think it does... I might be mistaken.   

Would you like me to add the profile and the `--profile, "kafka"`?
